contracts/Aavegotchi/ForgeDiamond/facets/ForgeFacet.sol

```diff
@@ -0,0 +1,461 @@
+    // @notice Will revert if aavegotchi is in active rental as well.
+    modifier onlyAavegotchiOwner(uint256 gotchiId) {
+        require(!lendingGetterAndSetterFacet().isAavegotchiLent(uint32(gotchiId)), "ForgeFacet: Aavegotchi is lent out");
+//        require(!lendingGetterAndSetterFacet().isAavegotchiListed(uint32(gotchiId)), "ForgeFacet: Aavegotchi is listed for rental");
+        require(LibMeta.msgSender() == aavegotchiFacet().ownerOf(gotchiId), "ForgeFacet: Not Aavegotchi owner");
+        _;
+    }
+
+    function forgeTokenFacet() internal view returns (ForgeTokenFacet facet){
+        facet = ForgeTokenFacet(address(this));
+    }
+    // External contracts
+    function aavegotchiGameFacet() internal pure returns (AavegotchiGameFacet facet){
+        facet = AavegotchiGameFacet(ForgeLibDiamond.AAVEGOTCHI_DIAMOND);
+    }
+    function aavegotchiFacet() internal pure returns (AavegotchiFacet facet){
+        facet = AavegotchiFacet(ForgeLibDiamond.AAVEGOTCHI_DIAMOND);
+    }
+    function itemsFacet() internal pure returns (ItemsFacet facet){
+        facet = ItemsFacet(ForgeLibDiamond.AAVEGOTCHI_DIAMOND);
+    }
+    function wearablesFacet() internal pure returns (WearablesFacet facet){
+        facet = WearablesFacet(ForgeLibDiamond.WEARABLE_DIAMOND);
+    }
+    function lendingGetterAndSetterFacet() internal pure returns (LendingGetterAndSetterFacet facet){
+        facet = LendingGetterAndSetterFacet(ForgeLibDiamond.AAVEGOTCHI_DIAMOND);
+    }
+    function gltrContract() internal returns (IERC20 token){
+        token = IERC20(s.GLTR);
+    }
+    ////////
+
+
+    // @notice Get an Aavegotchi's current smithing skill level
+    // @dev due to complex formula (approx P = 8 * 1.4^L for each next level), thresholds hardcoded here.
+    function getAavegotchiSmithingLevel(uint256 gotchiId) public view returns (uint256) {
+        uint256[30] memory sequence =
+        [uint256(0), 16, 38, 69, 113, 174, 259, 378, 544, 776, 1100, 1554, 2189, 3078, 4323,
+        6066, 8506, 11922, 16704, 23398, 32769, 45889, 64256, 89970, 125970, 176369, 246928,
+        345710, 484004, 677616];
+
+        uint256 points = s.gotchiSmithingSkillPoints[gotchiId];
+
+        if (points > sequence[sequence.length - 1]){
+            return sequence.length;
+        }
+
+        uint256 level = 0;
+        for (uint256 i = 0; i < sequence.length; i++) {
+            // minimum level is 1 (hence >=), i.e. 1.00x forge time multiplier.
+            if (points >= sequence[i]) {
+                level++;
+            } else {
+                break;
+            }
+        }
+        return level;
+    }
+
+    function getAavegotchiSmithingSkillPts(uint256 gotchiId) public view returns (uint256) {
+        return s.gotchiSmithingSkillPoints[gotchiId];
+    }
+
+
+    // @notice Return the forge time multiplier gained from smithing level, represented in bips.
+    function getSmithingLevelMultiplierBips(uint256 gotchiId) public view returns (uint256) {
```

Comment:
will add comment. i dont think we can avoid array somewhere, since to calculate i would need to keep track of how many items each gotchi has both forged and smelted separately, and on top of that, per rarity of each of those.

Reply:
i will add the comment.
you cant use the formula directly due to solidity not supporting decimals, and I'm having some trouble getting it to work correctly at intended precision using some digits offsets. please let me know if you have a solution in mind, otherwise I dont think doing it this way has much of a negative impact.